### PR TITLE
Refactor controlnet and add img2img and inpaint

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# ControlNet refactor and Img2Img and Inpaint

This PR is added upon frequent request from the community: https://github.com/huggingface/diffusers/issues/3095#issuecomment-1538004295

**NOTE:**  This PR is heavily influenced / copied from @haofanwang's https://github.com/haofanwang/ControlNet-for-Diffusers/ as well as @mikegarts' addition here: https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_controlnet_img2img.py 

A big thank you to both :heart: 

This PR makes sure that both scripts are tested and have all the new features added by @takuma104.

This PR moves ControlNet into its own directory as it's a highly used model. 
Also Img2Img and Inpainting is added in this PR.

Examples:
```py
# !pip install opencv-python transformers accelerate
from diffusers import StableDiffusionControlNetImg2ImgPipeline, ControlNetModel, UniPCMultistepScheduler
from diffusers.utils import load_image
import numpy as np
import torch

import cv2
from PIL import Image

# download an image
image = load_image(
    "https://hf.co/datasets/huggingface/documentation-images/resolve/main/diffusers/input_image_vermeer.png"
)
np_image = np.array(image)

# get canny image
np_image = cv2.Canny(np_image, 100, 200)
np_image = np_image[:, :, None]
np_image = np.concatenate([np_image, np_image, np_image], axis=2)
canny_image = Image.fromarray(np_image)

# load control net and stable diffusion v1-5
controlnet = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-canny", torch_dtype=torch.float16)
pipe = StableDiffusionControlNetImg2ImgPipeline.from_pretrained(
    "runwayml/stable-diffusion-v1-5", controlnet=controlnet, torch_dtype=torch.float16
)

# speed up diffusion process with faster scheduler and memory optimization
pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)
pipe.enable_model_cpu_offload()

# generate image
generator = torch.manual_seed(0)
image = pipe(
    "futuristic-looking woman",
    num_inference_steps=20,
    generator=generator,
    image=image,
    control_image=canny_image,
).images[0]
```
aa (5)

and

```py
# !pip install opencv-python transformers accelerate
from diffusers import StableDiffusionControlNetInpaintPipeline, ControlNetModel, UniPCMultistepScheduler
from diffusers.utils import load_image
import numpy as np
import torch

import cv2
from PIL import Image

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url).resize((512, 512))
mask_image = load_image(mask_url).resize((512, 512))

image = np.array(init_image)

# get canny image
image = cv2.Canny(image, 100, 200)
image = image[:, :, None]
image = np.concatenate([image, image, image], axis=2)
canny_image = Image.fromarray(image)

# load control net and stable diffusion inpainting
controlnet = ControlNetModel.from_pretrained("lllyasviel/sd-controlnet-canny", torch_dtype=torch.float16)
pipe = StableDiffusionControlNetInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting", controlnet=controlnet, torch_dtype=torch.float16
)

# speed up diffusion process with faster scheduler and memory optimization
pipe.scheduler = UniPCMultistepScheduler.from_config(pipe.scheduler.config)

pipe.enable_model_cpu_offload()

# generate image
generator = torch.manual_seed(0)
image = pipe(
    "a red monkey",
    num_inference_steps=30,
    generator=generator,
    image=init_image,
    mask_image=mask_image,
    control_image=canny_image,
).images[0]
```

spiderman

**Inpainting does seem to always work out of the box, it often needs some tweaking. This makes sense as the inpainting checkpoint was not used to train ControlNet**.

### CODE REVIEW

File: tests/pipelines/stable_diffusion/test_stable_diffusion_image_variation.py [hunk 1]
```diff
@@ -46,9 +46,7 @@ class StableDiffusionImageVariationPipelineFastTests(
     pipeline_class = StableDiffusionImageVariationPipeline
     params = IMAGE_VARIATION_PARAMS
     batch_params = IMAGE_VARIATION_BATCH_PARAMS
-    image_params = frozenset(
-        []
-    )  # TO-DO: update image_params once pipeline is refactored with VaeImageProcessor.preprocess
+    image_params = frozenset([])
```
Comment (patrickvonplaten) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_image_variation.py:49]:
```suggestion
    image_params = frozenset([])
    # TO-DO: update image_params once pipeline is refactored with VaeImageProcessor.preprocess
```

**Review by sayakpaul — APPROVED** [review]
LGTM. 

Some questions left in Slack for internal discussions.

**Comment by patrickvonplaten** [comment]
Left some more authorship, think we can merge it.

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py [hunk 2]
```diff
@@ -0,0 +1,1228 @@
...
+        generator: Optional[Union[torch.Generator, List[torch.Generator]]] = None,
+        latents: Optional[torch.FloatTensor] = None,
+        prompt_embeds: Optional[torch.FloatTensor] = None,
+        negative_prompt_embeds: Optional[torch.FloatTensor] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+        callback: Optional[Callable[[int, int, torch.FloatTensor], None]] = None,
+        callback_steps: int = 1,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        controlnet_conditioning_scale: Union[float, List[float]] = 0.5,
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_inpaint.py:930]: default is just 0.5 as 1 is too strong

File: tests/pipelines/controlnet/test_controlnet_inpaint.py [hunk 3]
```diff
@@ -13,6 +13,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# This model implementation is heavily based on:
```
Comment (sayakpaul) [inline tests/pipelines/controlnet/test_controlnet_inpaint.py:16]: @patrickvonplaten did you mean to add anything here? Seems incomplete to me.

**Pull request merged**